### PR TITLE
PoC: CockroachDB, Quarkus based keycloak and legacy store

### PR DESCRIPTION
**Pull request by DGuhr**
Draft PR combining @xgp and my changes from #10430 / #12400 / #12627 to make cockroach work with the legacy store and an integration test proving that it works in general.

Thanks @xgp for the help. 

@pedroigor @hmlnarik @stianst I guess this is a good point to stop for now until further decisions on the actual integration are made.

I have no idea if we should follow this path further or 100% go new-store and drop completely. Nevertheless I wanted proof it works (in general), so here's a draft PR with an integration test that proofs quarkus-based keycloak runs on legacy store with cockroachdb, can be spun up successfully and import/export stuff. (Couldn't test password authentication, as the container runs insecure cdb single-node instance, that insecure mode doesn't allow passwords at all.  

These changes would need a rebase/merge when https://github.com/keycloak/keycloak/pull/12627 (predecessor) is merged to main. 

Also there's surely potential for UX optimization (atm people would have to set transaction-jta-enabled=false manually, this could be checked automagically when they build with db=cockroach).

### CODE REVIEW

**Comment by xgp** [comment]
Hi @DGuhr, this looks great. Anything I can do to help push this across the finish line?

**Comment by xgp** [comment]
Hi @DGuhr Is there any plan to get this merged into main? Anything else I can do to help? Thanks!

**Comment by xgp** [comment]
If anyone is interested, [this branch](https://github.com/xgp/keycloak/tree/19.0.1_crdb) works with cockroach with the "legacy store", and I published a Docker image at `quay.io/phasetwo/keycloak-crdb:19.0.1`

**Comment by hmlnarik** [comment]
@DGuhr Thank you for the proof of concept.

TL;DR: CockroachDB will be supported [in the new store](https://www.keycloak.org/2022/07/storage-map). It is not expected to be maintained in the legacy store.

As for CRDB, community interest is clearly there. This interest has been answered in the new store where CockroachDB is going to be fully supported database. As the new store is getting close to production-ready state, we have to balance out the maintenance cost of two versus one implementation.

In the past, Keycloak has supported several databases in, support for which has been eventually dropped: DB2, MongoDB. Reason for removing their support was too high maintenance cost. The same argument applies when deciding one or two implementations - maintenance cost is nontrivial for adding another legacy store implementation.

Furthermore, legacy store relies on cooperation between infinispan and database. For data stored in the database, Infinispan is used as a cache, and relies on proper invalidation messages being sent across the clusters and sites. Limits to this approach have been found which will be addressed in the new store and not in the legacy store. Specifically, CockroachDB as a database spanning multiple sites will serve better without invalidation messages needed, and for that we need the new store.

Therefore, with appreciation of this PR efforts, I suggest closing this PR.

**Comment by xgp** [comment]
We've been running our CRDB fork in production for >18 months without issues, and have many large commercial customers running it under large load. We've repeatedly offered it to Keycloak, without response. Despite RedHat having a deal with Cockroach Labs, there no longer seems to be any appetite to support it.
- Branch of `24.0.3_crdb`: https://github.com/p2-inc/keycloak/tree/24.0.3_crdb
- Diff from Keycloak `24.0.3`: https://github.com/p2-inc/keycloak/commit/07888266cde76ab990358ae73484d92cdddc1960
- Docker image: https://quay.io/repository/phasetwo/keycloak-crdb `quay.io/phasetwo/keycloak-crdb`

**Comment by xgp** [comment]
FYI added an enhancement issue here https://github.com/keycloak/keycloak/issues/28931 where I will track the suggested PR.

**Comment by ahus1** [comment]
Please join the discussion in #8885 to make a case for CockroachDB. Making a case would include showing customer-facing benefits of CockroachDB over other PostgreSQL flavors.
